From a1bf3abe0c1093e81e5dd59b6d3b09b9ebb3a17d Mon Sep 17 00:00:00 2001
From: Matthias Reichl <hias@horus.com>
Date: Thu, 30 Dec 2021 14:28:37 +0100
Subject: [PATCH] drm/vc4: hdmi: Fix HDMI monitor detection in polled
 mode

When vc4_hdmi_connector_detect() was called in
connector_status_connected state it incorrectly cleared the
hdmi_monitor flag, leading to no audio on RPi3.

Fix this by clearing hdmi_monitor only when the hpd check
indicated no connection or if reading the edid failed.

Signed-off-by: Matthias Reichl <hias@horus.com>
---
 drivers/gpu/drm/vc4/vc4_hdmi.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

--- a/drivers/gpu/drm/vc4/vc4_hdmi.c
+++ b/drivers/gpu/drm/vc4/vc4_hdmi.c
@@ -243,7 +243,6 @@ vc4_hdmi_connector_detect(struct drm_con
 			connected = true;
 	}
 
-	vc4_hdmi->encoder.hdmi_monitor = false;
 	if (connected) {
 		if (connector->status != connector_status_connected) {
 			struct edid *edid = drm_get_edid(connector, vc4_hdmi->ddc);
@@ -252,6 +251,8 @@ vc4_hdmi_connector_detect(struct drm_con
 				cec_s_phys_addr_from_edid(vc4_hdmi->cec_adap, edid);
 				vc4_hdmi->encoder.hdmi_monitor = drm_detect_hdmi_monitor(edid);
 				kfree(edid);
+			} else {
+				vc4_hdmi->encoder.hdmi_monitor = false;
 			}
 		}
 
@@ -261,6 +262,8 @@ vc4_hdmi_connector_detect(struct drm_con
 		return connector_status_connected;
 	}
 
+	vc4_hdmi->encoder.hdmi_monitor = false;
+
 	cec_phys_addr_invalidate(vc4_hdmi->cec_adap);
 	pm_runtime_put(&vc4_hdmi->pdev->dev);
 	mutex_unlock(&vc4_hdmi->mutex);
